qiskit_experiments/curve_analysis/__init__.py

```diff
@@ -17,6 +17,403 @@
 
 .. currentmodule:: qiskit_experiments.curve_analysis
 
+Curve analysis provides the analysis base class for variety of experiments with 1-D parameter scan.
+Subclass can override several class attributes to define the behavior of the
+data formatting and fitting. Here we describe how code developer can create new curve fit
+analysis inheriting from the base class.
+
+
+Overview
+========
+
+The base class :class:`CurveAnalysis` supports multi-objective optimization on
+different set of experiment results, and you can also define multiple independent
+optimization tasks in the same class. The analysis is implemented with following data model.
+
+- Group: This is top level component of the fitting. If an analysis defines
+  multiple groups, it performs multiple independent optimization
+  and generates results for every optimization group.
+
+- Series: This is a collection of curves to form a multi-objective optimization task.
+  The fit entries in the same series share the fit parameters,
+  and multiple experimental results are simultaneously fit to generate a single fit result.
+
+- Curve: This is a single entry of analysis. Every curve may take unique filter keywords
+  to extract corresponding (x, y) data from the whole experimental results,
+  along with the callback function used for the curve fitting.
+
+To manage this structure, curve analysis provides a special dataclass :class:`SeriesDef`
+that represents a optimization configuration for a single curve data.
+Based on this information, the analysis automatically constructs a proper optimization logic.
+Thus one can avoid writing boilerplate code in various curve analyses
+and quickly write up the analysis code for a particular experiment.
+This analysis generates a set of :class:`~qiskit_experiments.framework.AnalysisResultData`
+entries with a single Matplotlib plot of the fit curves with raw data points.
+
+.. _curve_analysis_define_new:
+
+Defining new curves
+===================
+
+You can intuitively write definition of new curve, as shown below
+
+.. code-block:: python3
+
+    from qiskit_experiments.curve_analysis import SeriesDef, fit_function
+
+    SeriesDef(
+        fit_func=lambda x, p0, p1, p2: fit_function.exponential_decay(
+            x, amp=p0, lamb=p1, baseline=p2
+        ),
+        model_description="p0 * exp(-p1 * x) + p2",
+    )
+
+The minimum field you must fill with is the ``fit_func``, which is a callback function used
+with the optimization solver. Here you must call one of fit functions from the module
+:mod:`qiskit_experiments.curve_analysis.fit_function` because they implement
+a special logic to compute error propagation.
```

Comment:
Not really. A function there is wrapped by this decorator and uses `unumpy` functions instead of `numpy` to compute error propagation. If one can do the same, every function is acceptable. Re-adding support for non-unumpy functions requires many branching logic everywhere, but I think we can still do that -- if you think this is useful :)
https://github.com/Qiskit/qiskit-experiments/blob/be322e3a8d709613716232bec4aabbaeb1ec97b3/qiskit_experiments/curve_analysis/fit_function.py#L25-L45